src/array.rs

```diff
@@ -0,0 +1,456 @@
+        Array { ptr }
+    }
+
+    /// New array from existing buffer.
+    ///
+    /// # Parameters
+    ///
+    /// - `data`: A buffer which will be copied.
+    /// - `shape`: Shape of the array.
+    /// - `dim`: Number of dimensions (size of shape).
+    pub fn from_data<T: ArrayElement>(data: &[T], shape: &[i32], dim: i32) -> Self {
+        let ptr = unsafe {
+            mlx_sys::mlx_array_from_data(
+                data.as_ptr() as *const c_void,
+                shape.as_ptr(),
+                dim,
+                T::DTYPE as u32,
+            )
+        };
+
+        Array { ptr }
+    }
+
+    /// The size of the array’s datatype in bytes.
+    pub fn itemsize(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_itemsize(self.ptr) }
+    }
+
+    /// Number of elements in the array.
+    pub fn size(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_size(self.ptr) }
+    }
+
+    /// The strides of the array.
+    pub fn strides(&self) -> &[usize] {
+        let ndim = self.ndim();
+
+        unsafe {
+            let data = mlx_sys::mlx_array_strides(self.ptr);
+            std::slice::from_raw_parts(data, ndim)
+        }
+    }
+
+    /// The number of bytes in the array.
+    pub fn nbytes(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_nbytes(self.ptr) }
+    }
+
+    /// The array’s dimension.
+    pub fn ndim(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_ndim(self.ptr) }
+    }
+
+    /// The shape of the array.
+    ///
+    /// Returns: a pointer to the sizes of each dimension.
+    pub fn shape(&self) -> &[i32] {
+        let ndim = self.ndim();
+
+        unsafe {
+            let data = mlx_sys::mlx_array_shape(self.ptr);
+            std::slice::from_raw_parts(data, ndim)
+        }
+    }
+
+    /// The shape of the array in a particular dimension.
+    ///
+    /// # Panic
+    ///
+    /// Panics if the array is scalar.
+    pub fn dim(&self, dim: i32) -> i32 {
+        // This will panic on a scalar array
+        unsafe { mlx_sys::mlx_array_dim(self.ptr, dim) }
+    }
+
+    /// The array element type.
+    pub fn dtype(&self) -> Dtype {
+        let dtype = unsafe { mlx_sys::mlx_array_get_dtype(self.ptr) };
+        Dtype::try_from(dtype).unwrap()
+    }
+
+    /// Evaluate the array.
+    pub fn eval(&mut self) {
+        // This clearly modifies the array, so it should be mutable
+        unsafe { mlx_sys::mlx_array_eval(self.ptr) };
+    }
+
+    /// Access the value of a scalar array.
+    pub fn item<T: ArrayElement>(&self) -> T {
+        T::scalar_array_item(self)
+    }
+
+    /// Returns a pointer to the array data
+    pub fn data<T: ArrayElement>(&self) -> &[T] {
+        // TODO: check below after ops are implemented
+        // Array must be evaluated, otherwise returns NULL.
+
+        let data = T::array_data(self);
+        let size = self.size();
+        unsafe { std::slice::from_raw_parts(data, size) }
+    }
+}
+
+impl From<bool> for Array {
```

Comment:
This might be a dumb suggestion but should we introduce a trait for values that can be promoted to an Array? This would work for scalars too. types that conform to ArrayElement would also conform to this new trait. Might not be super Rust-y though, so just a thought.

Reply:
Which do you think would be better? A new trait `ArrayElementExt` or just `impl From<T: ArrayElement> for Array`?

Reply:
My thinking was that if we have a new array that can be promoted to array, we could have some of the methods take that trait, and then they could give scalar values without themselves needing to promote it. We could just do that internally. Does that work even if we just do `impl From<T: ArrayElement> for Array`?

Reply:
I am not sure if I understand this correctly. Is the example below depicting what you were thinking?

```rust
let arr = Array::from_int(1);
let result: Array = add(arr, 1); // 1i32 can be converted into an `Array`
```

Reply:
Yes, this was my idea. But we can also leave it for now, and adjust as we actually start adding ops and start to get a feel for where the API could feel better

Reply:
I think `From<T>` would suffice. We would just need to do something like this when we impl ops

```rust
fn add(arr1: impl Into<Array>, arr2: impl Into<Array>, s: StreamOrDevice) -> Array {
    let arr1 = arr1.into();
    let arr2 = arr2.into();
    let c_array = unsafe { mlx_add(arr1.c_array, arr2.c_array, s) };
    Array { c_array }
}
```